logserver/src/main/java/com/yahoo/logserver/handlers/archive/FilesArchived.java

```diff
@@ -0,0 +1,214 @@
+ *
+ * @author Arne Juul
+ */
+public class FilesArchived {
+    private static final Logger log = Logger.getLogger(FilesArchived.class.getName());
+
+    /**
+     * File instance representing root directory of archive
+     */
+    private final File root;
+
+    // known-existing files inside the archive directory
+    private List<LogFile> knownFiles;
+
+    public final static long compressAfterMillis = 2L * 3600 * 1000;
+    private long maxAgeDays = 30; // GDPR rules: max 30 days
+    private long sizeLimit = 30L * (1L << 30); // 30 GB
+
+    /**
+     * Creates an FilesArchive managing the given directory
+     */
+    public FilesArchived(File rootDir) {
+        this.root = rootDir;
+        maintenance();
+    }
+
+    public String toString() {
+        return FilesArchived.class.getName() + ": root=" + root;
+    }
+
+    public int highestGen(String prefix) {
+        int gen = 0;
+        for (LogFile lf : knownFiles) {
+            if (prefix.equals(lf.prefix)) {
+                gen = Math.max(gen, lf.generation);
+            }
+        }
+        return gen;
+    }
+
+    public synchronized void maintenance() {
+        rescan();
+        if (removeOlderThan(maxAgeDays)) rescan();
+        if (compressOldFiles()) rescan();
+        long days = maxAgeDays;
+        while (tooMuchDiskUsage() && (--days > 1)) {
+            if (removeOlderThan(days)) rescan();
+        }
+    }
+
+    private void rescan() {
+        knownFiles = scanDir(root);
+    }
+
+    boolean tooMuchDiskUsage() {
+        long sz = sumFileSizes();
+        return sz > sizeLimit;
+    }
+
+    private boolean olderThan(LogFile lf, long days, long now) {
+        long mtime = lf.path.lastModified();
+        long diff = now - mtime;
+        return (diff > days * 86400L * 1000L);
+    }
+
+    // returns true if any files were removed
+    private boolean removeOlderThan(long days) {
+        boolean action = false;
+        long now = System.currentTimeMillis();
+        for (LogFile lf : knownFiles) {
+            if (olderThan(lf, days, now)) {
+                lf.path.delete();
+                log.info("Deleted: "+lf.path);
+                action = true;
+            }
+        }
+        return action;
+    }
+
+    // returns true if any files were compressed
+    private boolean compressOldFiles() {
+        boolean action = false;
+        long now = System.currentTimeMillis();
+        int count = 0;
+        for (LogFile lf : knownFiles) {
+            // avoid compressing entire archive at once
+            if (lf.canCompress(now) && (count++ < 5)) {
+                compress(lf.path);
+            }
+        }
+        return count > 0;
+    }
+
+    private void compress(File oldFile) {
+        File gzippedFile = new File(oldFile.getPath() + ".gz");
+        try {
+            long mtime = oldFile.lastModified();
+            GZIPOutputStream compressor = new GZIPOutputStream(new FileOutputStream(gzippedFile), 0x100000);
```

Comment:
Use "try-with-resource" to ensure all streams are closed if an exception is thrown during read/write.